src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
     ["SELECT id from factsets WHERE certname = ?" certname]
     (:id (first result-set))))
 
-(pls/defn-validated factset-map :- {s/Str s/Str}
-  "Return all facts and their values for a given certname as a map"
+(defn-validated delete-pending-path-id-orphans!
+  "Delete paths in removed-pid-vid-pairs that are no longer mentioned
+  in facts."
+  [factset-id removed-pid-vid-pairs]
+  (when-let [removed-pid-vid-pairs (seq removed-pid-vid-pairs)]
+    (let [pids (map first removed-pid-vid-pairs)
+          fv-id-pairs (map vector
+                           (repeat factset-id )
+                           (map second removed-pid-vid-pairs))
+          in-pids (jdbc/in-clause pids)
+          in-fv-id-pairs (jdbc/in-clause-multi fv-id-pairs 2)]
+      (sql/do-prepared
+       (format
+        "DELETE FROM fact_paths fp
+           WHERE fp.id %s
+             AND NOT EXISTS (SELECT 1 FROM facts f
+                               WHERE f.fact_path_id %s
+                                 AND f.fact_path_id = fp.id
+                                 AND (f.factset_id, f.fact_value_id) NOT %s)"
```

Comment:
@kbarber We didn't need the factset/value id pairs for fact_paths (since paths are unique per factset), but we do need the similar constraint for fact_values to protect different paths in the same factset that point to the same value, and that aren't being deleted in the current transaction.